include/parameter.h

```diff
@@ -0,0 +1,307 @@
+    virtual void set_value(T value);
+    virtual void set_value_transformed(T value_transformed);
+    virtual void set_unit(std::shared_ptr<const Unit> unit = nullptr);
+    virtual std::string str() const;
+
+
+    virtual ~ParameterBase() = default;
+};
+
+template<typename T, class C, class Unit>
+class Parameter : public ParameterBase<T, Unit>, public std::enable_shared_from_this<C> {
+public:
+    using SetC = std::set<std::shared_ptr<C>>;
+
+private:
+    // TODO: Remember why this class needs to exist, if it does
+    // Probably just to store the const Limits reference?
+    struct Limiter {
+        const Limits<T> & limits;
+        Limiter(const Limits<T> & limits_in) : limits(limits_in) {};
+    };
+
+    struct Transformer {
+        const Transform<T> & transform;
+        Transformer(const Transform<T> & transform_in) : transform(transform_in) {};
+    };
+
+    static inline constexpr T _default = 0;
+    static inline constexpr T _min = -std::numeric_limits<T>::infinity();
+    static inline constexpr T _max = std::numeric_limits<T>::infinity();
+
+    static inline const std::string _desc = "";
+    static inline const std::string _name = "";
+
+    template<typename P> static constexpr T _default_t(decltype(P::_default) *) { return P::_default; }
+    template<typename P> static constexpr T _default_t(...) { return _default; }
+    template<typename P> static constexpr T _min_t(decltype(P::_min) *) { return P::_min; }
+    template<typename P> static constexpr T _min_t(...) { return _min; }
+    template<typename P> static constexpr T _max_t(decltype(P::_max) *) { return P::_max; }
+    template<typename P> static constexpr T _max_t(...) { return _max; }
+    template<typename P> static const std::string _desc_t(decltype(P::_desc) *) { return P::_desc; }
+    template<typename P> static const std::string _desc_t(...) { return _desc; }
+    template<typename P> static const std::string _name_t(decltype(P::_name) *) { return P::_name; }
+    template<typename P> static const std::string _name_t(...) { return _name; }
+
+    bool _free = true;
+    std::unique_ptr<Limiter> _limiter;
+    std::unique_ptr<Transformer> _transformer;
+
+    std::string _label;
+    std::unique_ptr<SetC> _inheritors = nullptr;
+    std::unique_ptr<SetC> _modifiers = nullptr;
+
+    std::shared_ptr<const Limits<T>> _limits_ptr;
+    std::shared_ptr<const Transform<T>> _transform_ptr;
+    std::shared_ptr<const Unit> _unit_ptr;
+    std::shared_ptr<const C> _inheritee_ptr = nullptr;
+
+    void _set_value(T value)
+    {
+        if (!(get_limits().check(value))) throw std::runtime_error(
+            "Value=" + std::to_string(value) + "beyond get_limits()=" + get_limits().str()
+        );
+        _value = value;
+    }
+
+protected:
+    T _value;
+    T _value_transformed;
+
+    void set_inheritee(std::shared_ptr<const C> inheritee) {
+        _inheritee_ptr = inheritee == nullptr ? nullptr : std::move(inheritee);
```

Comment:
```suggestion
        _inheritee_ptr = inheritee ? nullptr : std::move(inheritee);
```
What you have here actually matches DM conventions, but I regret those because the rest of the C++ world [has gone in the other direction](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-if), and this code isn't beholden to DM conventions.

Reply:
Point taken, but are the CppCoreGuidelines really used universally, or reflect most of the rest of the C++ world?

Reply:
I'm sure they're far from universal; too many existing guidelines with big codebase behind them.  But I do expect them to be what all of those other guidelines slowly evolve toward, and more importantly they've been thought about harder by cumulatively much more experienced people than our own guidelines.